automation/patch_igz/patch_remote.py

```diff
@@ -0,0 +1,356 @@
+                    self._registry_username,
+                    "--password",
+                    self._registry_password,
+                ],
+                live=True,
+            )
+
+        self._push_docker_image(built_image)
+
+        node = nodes[0]
+        self._connect_to_node(node)
+        try:
+            self._replace_deploy_policy()
+            self._replace_deployment_images(built_image)
+            self._wait_deployment_ready()
+        finally:
+            out = self._exec_remote(
+                [
+                    "kubectl",
+                    "-n",
+                    "default-tenant",
+                    "get",
+                    "pods",
+                ],
+            )
+
+            for line in out.splitlines():
+                if "mlrun-api" in line:
+                    logger.info(line)
+
+            self._disconnect_from_node()
+
+        logger.info(
+            "Deployed branch successfully! Yay! (Note this may not survive system restarts)"
+        )
+
+    def _validate_config(self):
+        missing_fields = self.Consts.mandatory_fields - set(self._config.keys())
+        if len(missing_fields) > 0:
+            raise RuntimeError(f"Mandatory options not defined: {missing_fields}")
+
+        self._registry_username = self._config.get("REGISTRY_USERNAME")
+        self._registry_password = self._config.get("REGISTRY_PASSWORD")
+        if self._registry_username is not None and self._registry_password is None:
+            raise RuntimeError(
+                "REGISTRY_USERNAME defined, yet REGISTRY_PASSWORD is not defined"
+            )
+
+    @contextmanager
+    def _add_mlrun_src_to_path(self):
+        mlrun_src = os.path.dirname(os.path.abspath(__file__)) + "/.."
+        sys.path.append(mlrun_src)
+        yield sys.path
+        sys.path.remove(mlrun_src)
+
+    def _get_current_version(self) -> str:
+        with self._add_mlrun_src_to_path():
+            from version import version_file
+
+            return str(version_file.read_unstable_version_prefix())
+
+    def _find_current_git_branch(self) -> str:
+        cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"]
+        return self._exec_local(cmd).strip()
+
+    def _make_mlrun_api(self, image_tag):
+        logger.info("Building mlrun-api docker image")
+        os.environ["MLRUN_VERSION"] = image_tag
+        os.environ["MLRUN_DOCKER_REPO"] = self._config["DOCKER_REGISTRY"]
+        cmd = ["make", "api"]
+        return self._exec_local(cmd, live=True).strip()
+
+    def _connect_to_node(self, node):
+        logger.debug(f"Connecting to {node}")
+
+        self._ssh_client = paramiko.SSHClient()
+        self._ssh_client.set_missing_host_key_policy(paramiko.WarningPolicy)
+        self._ssh_client.connect(
+            node,
+            username=self._config["SSH_USER"],
+            password=self._config["SSH_PASSWORD"],
+        )
+
+    def _disconnect_from_node(self):
+        self._ssh_client.close()
```

Comment:
Also here you can do a context manager (also not necessary)
```suggestion
    @contextmanager
    def connect_to_data_node(self, node):
        logger.debug(f"Connecting to {node}")

        ssh_client = paramiko.SSHClient()
        ssh_client.set_missing_host_key_policy(paramiko.WarningPolicy)
        ssh_client.connect(
            node,
            username=self._config["SSH_USER"],
            password=self._config["SSH_PASSWORD"],
        )
        yield ssh_client
        ssh_client.close()
```

And then you can:
```python
with self. connect_to_data_node(node) as client:
    # do stuff with client object
```

Reply:
This will not save anything, as I want to do stuff in "finally" anyway